File: android/sdk/src/main/kotlin/com/clevertap/android/nativedisplay/models/Layout.kt
package com.clevertap.android.nativedisplay.models

import androidx.compose.runtime.Immutable
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.nullable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.float
import kotlinx.serialization.json.jsonPrimitive

/**
 * Represents a dimension value that can be either a specific value or a special dimension.
 */
@Immutable
@Serializable
data class Dimension(
    val value: Float = 0f,
    val unit: DimensionUnit = DimensionUnit.DP,
    val special: SpecialDimension? = null
) {
    companion object {
        fun dp(value: Float) = Dimension(value, DimensionUnit.DP)
        fun sp(value: Float) = Dimension(value, DimensionUnit.SP)
        fun percent(value: Float) = Dimension(value, DimensionUnit.PERCENT)
        fun px(value: Float) = Dimension(value, DimensionUnit.PX)

        val WRAP_CONTENT = Dimension(0f, DimensionUnit.DP, SpecialDimension.WRAP_CONTENT)
        val MATCH_PARENT = Dimension(0f, DimensionUnit.DP, SpecialDimension.MATCH_PARENT)
    }
}

/**
 * Serializer for Dimension that accepts both raw numbers (treated as DP) and
 * object format `{"value": N, "unit": "percent"}`.
 *
 * Used by `borderRadius` in Style so that:
 * - `"borderRadius": 12` → `Dimension(12f, DimensionUnit.DP)`
 * - `"borderRadius": {"value": 50, "unit": "percent"}` → `Dimension(50f, DimensionUnit.PERCENT)`
 */
object DimensionAsNumberSerializer : KSerializer<Dimension> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("DimensionAsNumber")

    override fun deserialize(decoder: Decoder): Dimension {
        val jsonDecoder = decoder as JsonDecoder
        val element = jsonDecoder.decodeJsonElement()
        return when {
            element is JsonPrimitive && !element.isString -> {
                // Raw number → DP (backward compatible)
                Dimension(value = element.float, unit = DimensionUnit.DP)
            }
            element is JsonObject -> {
                // Object format: {"value": N, "unit": "percent"|"dp"|...}
                val value = element["value"]?.jsonPrimitive?.float ?: 0f
                val unitStr = element["unit"]?.jsonPrimitive?.content ?: "dp"
                val unit = when (unitStr.lowercase()) {
                    "percent" -> DimensionUnit.PERCENT
                    "sp" -> DimensionUnit.SP
                    "px" -> DimensionUnit.PX
                    else -> DimensionUnit.DP
                }
                Dimension(value = value, unit = unit)
            }
            else -> Dimension(value = 0f, unit = DimensionUnit.DP)
        }
    }

    override fun serialize(encoder: Encoder, value: Dimension) {
        val jsonEncoder = encoder as JsonEncoder
        if (value.unit == DimensionUnit.DP && value.special == null) {
            // Serialize simple DP values as raw numbers for compact JSON output
            jsonEncoder.encodeJsonElement(JsonPrimitive(value.value))
        } else {
            jsonEncoder.encodeJsonElement(
                JsonObject(
                    mapOf(
                        "value" to JsonPrimitive(value.value),
                        "unit" to JsonPrimitive(value.unit.name.lowercase())
                    )
                )
            )
        }
    }
}

/**
 * Nullable wrapper around [DimensionAsNumberSerializer].
 * Used by `Style.borderRadius` so that a JSON `null` deserializes to Kotlin `null`.
 */
object DimensionAsNumberSerializerNullable : KSerializer<Dimension?> {
    private val inner = DimensionAsNumberSerializer.nullable
    override val descriptor: SerialDescriptor = inner.descriptor

    override fun deserialize(decoder: Decoder): Dimension? = inner.deserialize(decoder)

    override fun serialize(encoder: Encoder, value: Dimension?) = inner.serialize(encoder, value)
}

/**
 * Offset for absolute positioning within a container (x, y coordinates).
 * Used for positioning elements at specific locations within Box containers.
 * Supports negative values for positioning outside the normal flow.
 */
@Immutable
@Serializable
data class Offset(
    val x: Float = 0f,
    val y: Float = 0f,
    val unit: DimensionUnit = DimensionUnit.DP
) {
    companion object {
        fun dp(x: Float, y: Float) = Offset(x, y, DimensionUnit.DP)
        fun percent(x: Float, y: Float) = Offset(x, y, DimensionUnit.PERCENT)
        
        val ZERO = Offset(0f, 0f)
    }
}

/**
 * Spacing values for padding.
 * Can specify individual sides or use shortcuts for all/horizontal/vertical.
 */
@Immutable
@Serializable
data class Spacing(
    val all: Float? = null,
    val horizontal: Float? = null,
    val vertical: Float? = null,
    val top: Float? = null,
    val bottom: Float? = null,
    val left: Float? = null,
    val right: Float? = null,
    val unit: DimensionUnit = DimensionUnit.DP
) {
    /**
     * Resolve individual spacing values with proper fallbacks.
     */
    fun resolveTop(): Float = top ?: vertical ?: all ?: 0f
    fun resolveBottom(): Float = bottom ?: vertical ?: all ?: 0f
    fun resolveLeft(): Float = left ?: horizontal ?: all ?: 0f
    fun resolveRight(): Float = right ?: horizontal ?: all ?: 0f
    
    companion object {
        fun all(value: Float, unit: DimensionUnit = DimensionUnit.DP) = 
            Spacing(all = value, unit = unit)
        
        fun horizontal(value: Float, unit: DimensionUnit = DimensionUnit.DP) = 
            Spacing(horizontal = value, unit = unit)
        
        fun vertical(value: Float, unit: DimensionUnit = DimensionUnit.DP) = 
            Spacing(vertical = value, unit = unit)
    }
}

/**
 * Child arrangement strategy for container layouts.
 * Defines how children are positioned and spaced within Column/Row containers.
 */
@Immutable
@Serializable
data class ChildArrangement(
    /**
     * Spacing value between children (used when strategy is SPACED).
     */
    val spacing: Float? = null,
    
    /**
     * Unit for spacing value.
     */
    val spacingUnit: DimensionUnit = DimensionUnit.DP,
    
    /**
     * Arrangement strategy for positioning children.
     */
    val strategy: ArrangementStrategy = ArrangementStrategy.SPACED
) {
    companion object {
        /**
         * Default arrangement: no spacing between children.
         */
        val DEFAULT = ChildArrangement(spacing = 0f, strategy = ArrangementStrategy.SPACED)
        
        /**
         * Fixed spacing between children.
         */
        fun spaced(spacing: Float, unit: DimensionUnit = DimensionUnit.DP) = 
            ChildArrangement(spacing = spacing, spacingUnit = unit, strategy = ArrangementStrategy.SPACED)
        
        /**
         * Equal space between children, no space at edges.
         */
        fun spaceBetween() = 
            ChildArrangement(strategy = ArrangementStrategy.SPACE_BETWEEN)
        
        /**
         * Equal space between children AND at edges.
         */
        fun spaceEvenly() = 
            ChildArrangement(strategy = ArrangementStrategy.SPACE_EVENLY)
        
        /**
         * Equal space around each child (half space at edges).
         */
        fun spaceAround() = 
            ChildArrangement(strategy = ArrangementStrategy.SPACE_AROUND)
        
        /**
         * Children aligned to start, no spacing.
         */
        fun start() = 
            ChildArrangement(strategy = ArrangementStrategy.START)
        
        /**
         * Children centered, no spacing.
         */
        fun center() = 
            ChildArrangement(strategy = ArrangementStrategy.CENTER)
        
        /**
         * Children aligned to end, no spacing.
         */
        fun end() = 
            ChildArrangement(strategy = ArrangementStrategy.END)
    }
}

/**
 * Layout properties for positioning and sizing elements.
 */
@Immutable
@Serializable
data class Layout(
    val width: Dimension? = null,
    val height: Dimension? = null,
    val aspectRatio: Float? = null,  // Maintain fixed width:height ratio (e.g., 1.777 for 16:9)
    val offset: Offset? = null,  // Changed from margin: Spacing to offset: Offset
    val padding: Spacing? = null,
    val arrangement: ChildArrangement? = null  // Changed from spacing/spacingUnit to arrangement
) {
    companion object {
        val DEFAULT = Layout(
            width = Dimension.WRAP_CONTENT,
            height = Dimension.WRAP_CONTENT
        )
    }
}

File: ios/Sources/CleverTapNativeDisplay/Models/Layout.swift
// MARK: - Layout Models
// Layout and dimension configuration for the Native Display System

import Foundation

/// Represents a dimension value that can be either a specific value or a special dimension.
public struct Dimension: Codable, Equatable {
    public let value: CGFloat
    public let unit: DimensionUnit
    public let special: SpecialDimension?
    
    public init(
        value: CGFloat = 0,
        unit: DimensionUnit = .dp,
        special: SpecialDimension? = nil
    ) {
        self.value = value
        self.unit = unit
        self.special = special
    }
    
    // Custom decoder to handle defaults
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.value = try container.decodeIfPresent(CGFloat.self, forKey: .value) ?? 0
        self.unit = try container.decodeIfPresent(DimensionUnit.self, forKey: .unit) ?? .dp
        self.special = try container.decodeIfPresent(SpecialDimension.self, forKey: .special)
    }
    
    private enum CodingKeys: String, CodingKey {
        case value, unit, special
    }
    
    // MARK: - Factory Methods
    
    public static func dp(_ value: CGFloat) -> Dimension {
        Dimension(value: value, unit: .dp)
    }
    
    public static func sp(_ value: CGFloat) -> Dimension {
        Dimension(value: value, unit: .sp)
    }
    
    public static func percent(_ value: CGFloat) -> Dimension {
        Dimension(value: value, unit: .percent)
    }
    
    public static func px(_ value: CGFloat) -> Dimension {
        Dimension(value: value, unit: .px)
    }
    
    public static let wrapContent = Dimension(value: 0, unit: .dp, special: .wrapContent)
    public static let matchParent = Dimension(value: 0, unit: .dp, special: .matchParent)
}

/// Offset for absolute positioning within a container (x, y coordinates).
/// Used for positioning elements at specific locations within ZStack containers.
/// Supports negative values for positioning outside the normal flow.
public struct Offset: Codable, Equatable {
    public let x: CGFloat
    public let y: CGFloat
    public let unit: DimensionUnit
    
    public init(
        x: CGFloat = 0,
        y: CGFloat = 0,
        unit: DimensionUnit = .dp
    ) {
        self.x = x
        self.y = y
        self.unit = unit
    }
    
    // Custom decoder to handle defaults
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.x = try container.decodeIfPresent(CGFloat.self, forKey: .x) ?? 0
        self.y = try container.decodeIfPresent(CGFloat.self, forKey: .y) ?? 0
        self.unit = try container.decodeIfPresent(DimensionUnit.self, forKey: .unit) ?? .dp
    }
    
    private enum CodingKeys: String, CodingKey {
        case x, y, unit
    }
    
    // MARK: - Factory Methods
    
    public static func dp(x: CGFloat, y: CGFloat) -> Offset {
        Offset(x: x, y: y, unit: .dp)
    }
    
    public static func percent(x: CGFloat, y: CGFloat) -> Offset {
        Offset(x: x, y: y, unit: .percent)
    }
    
    public static let zero = Offset(x: 0, y: 0)
}

/// Spacing values for padding.
/// Can specify individual sides or use shortcuts for all/horizontal/vertical.
public struct Spacing: Codable, Equatable {
    public let all: CGFloat?
    public let horizontal: CGFloat?
    public let vertical: CGFloat?
    public let top: CGFloat?
    public let bottom: CGFloat?
    public let left: CGFloat?
    public let right: CGFloat?
    public let unit: DimensionUnit
    
    public init(
        all: CGFloat? = nil,
        horizontal: CGFloat? = nil,
        vertical: CGFloat? = nil,
        top: CGFloat? = nil,
        bottom: CGFloat? = nil,
        left: CGFloat? = nil,
        right: CGFloat? = nil,
        unit: DimensionUnit = .dp
    ) {
        self.all = all
        self.horizontal = horizontal
        self.vertical = vertical
        self.top = top
        self.bottom = bottom
        self.left = left
        self.right = right
        self.unit = unit
    }
    
    // Custom decoder to handle defaults
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.all = try container.decodeIfPresent(CGFloat.self, forKey: .all)
        self.horizontal = try container.decodeIfPresent(CGFloat.self, forKey: .horizontal)
        self.vertical = try container.decodeIfPresent(CGFloat.self, forKey: .vertical)
        self.top = try container.decodeIfPresent(CGFloat.self, forKey: .top)
        self.bottom = try container.decodeIfPresent(CGFloat.self, forKey: .bottom)
        self.left = try container.decodeIfPresent(CGFloat.self, forKey: .left)
        self.right = try container.decodeIfPresent(CGFloat.self, forKey: .right)
        self.unit = try container.decodeIfPresent(DimensionUnit.self, forKey: .unit) ?? .dp
    }
    
    private enum CodingKeys: String, CodingKey {
        case all, horizontal, vertical, top, bottom, left, right, unit
    }
    
    // MARK: - Resolution Methods
    
    /// Resolve individual spacing values with proper fallbacks.
    public func resolveTop() -> CGFloat {
        top ?? vertical ?? all ?? 0
    }
    
    public func resolveBottom() -> CGFloat {
        bottom ?? vertical ?? all ?? 0
    }
    
    public func resolveLeft() -> CGFloat {
        left ?? horizontal ?? all ?? 0
    }
    
    public func resolveRight() -> CGFloat {
        right ?? horizontal ?? all ?? 0
    }
    
    // MARK: - Factory Methods
    
    public static func all(_ value: CGFloat, unit: DimensionUnit = .dp) -> Spacing {
        Spacing(all: value, unit: unit)
    }
    
    public static func horizontal(_ value: CGFloat, unit: DimensionUnit = .dp) -> Spacing {
        Spacing(horizontal: value, unit: unit)
    }
    
    public static func vertical(_ value: CGFloat, unit: DimensionUnit = .dp) -> Spacing {
        Spacing(vertical: value, unit: unit)
    }
    
    public static let zero = Spacing()
}

/// Child arrangement strategy for container layouts.
/// Defines how children are positioned and spaced within VStack/HStack containers.
public struct ChildArrangement: Codable, Equatable {
    /// Spacing value between children (used when strategy is .spaced).
    public let spacing: CGFloat?
    
    /// Unit for spacing value.
    public let spacingUnit: DimensionUnit
    
    /// Arrangement strategy for positioning children.
    public let strategy: ArrangementStrategy
    
    public init(
        spacing: CGFloat? = nil,
        spacingUnit: DimensionUnit = .dp,
        strategy: ArrangementStrategy = .spaced
    ) {
        self.spacing = spacing
        self.spacingUnit = spacingUnit
        self.strategy = strategy
    }
    
    // Custom decoder to handle defaults
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.spacing = try container.decodeIfPresent(CGFloat.self, forKey: .spacing)
        self.spacingUnit = try container.decodeIfPresent(DimensionUnit.self, forKey: .spacingUnit) ?? .dp
        self.strategy = try container.decodeIfPresent(ArrangementStrategy.self, forKey: .strategy) ?? .spaced
    }
    
    private enum CodingKeys: String, CodingKey {
        case spacing, spacingUnit, strategy
    }
    
    // MARK: - Factory Methods
    
    /// Default arrangement: no spacing between children.
    public static let `default` = ChildArrangement(spacing: 0, strategy: .spaced)
    
    /// Fixed spacing between children.
    public static func spaced(_ spacing: CGFloat, unit: DimensionUnit = .dp) -> ChildArrangement {
        ChildArrangement(spacing: spacing, spacingUnit: unit, strategy: .spaced)
    }
    
    /// Equal space between children, no space at edges.
    public static func spaceBetween() -> ChildArrangement {
        ChildArrangement(strategy: .spaceBetween)
    }
    
    /// Equal space between children AND at edges.
    public static func spaceEvenly() -> ChildArrangement {
        ChildArrangement(strategy: .spaceEvenly)
    }
    
    /// Equal space around each child (half space at edges).
    public static func spaceAround() -> ChildArrangement {
        ChildArrangement(strategy: .spaceAround)
    }
    
    /// Children aligned to start, no spacing.
    public static func start() -> ChildArrangement {
        ChildArrangement(strategy: .start)
    }
    
    /// Children centered, no spacing.
    public static func center() -> ChildArrangement {
        ChildArrangement(strategy: .center)
    }
    
    /// Children aligned to end, no spacing.
    public static func end() -> ChildArrangement {
        ChildArrangement(strategy: .end)
    }
}

/// Layout properties for positioning and sizing elements.
public struct Layout: Codable, Equatable {
    public let width: Dimension?
    public let height: Dimension?
    public let offset: Offset?  // Changed from margin
    public let padding: Spacing?
    public let arrangement: ChildArrangement?  // Changed from spacing/spacingUnit
    
    public init(
        width: Dimension? = nil,
        height: Dimension? = nil,
        offset: Offset? = nil,
        padding: Spacing? = nil,
        arrangement: ChildArrangement? = nil
    ) {
        self.width = width
        self.height = height
        self.offset = offset
        self.padding = padding
        self.arrangement = arrangement
    }
    
    // Custom decoder to handle defaults
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.width = try container.decodeIfPresent(Dimension.self, forKey: .width)
        self.height = try container.decodeIfPresent(Dimension.self, forKey: .height)
        self.offset = try container.decodeIfPresent(Offset.self, forKey: .offset)
        self.padding = try container.decodeIfPresent(Spacing.self, forKey: .padding)
        self.arrangement = try container.decodeIfPresent(ChildArrangement.self, forKey: .arrangement)
    }
    
    private enum CodingKeys: String, CodingKey {
        case width, height, offset, padding, arrangement
    }
    
    public static let `default` = Layout(
        width: .wrapContent,
        height: .wrapContent
    )
}
